{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Map S-sulphenylated peptide fragments on 3D Structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The goal of this study is to systematically map the positions of S-sulphenylation of proteins onto 3D protein structures in the Protein Data Bank."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data source:\n",
    "\n",
    "Gupta V, Yang J, Liebler DC, Carroll KS. Diverse Redoxome Reactivity Profiles \n",
    "of Carbon Nucleophiles. J Am Chem Soc. 2017 Apr 19;139(15):5588-5595. [doi:\n",
    "10.1021/jacs.7b01791](https://doi.org/10.1021/jacs.7b01791)\n",
    "\n",
    "Excerpt from abstract:\n",
    "Analysis of sulfenic acid-reactive C-nucleophile fragments screened against a colon cancer cell proteome. Covalent ligands were identified for >1280 S-sulfenylated cysteines present in \"druggable\" proteins and orphan targets, revealing disparate reactivity profiles and target preferences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from io import BytesIO\n",
    "import xlrd\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import py3Dmol\n",
    "from pyspark.sql import SparkSession\n",
    "from pyspark.sql.functions import explode, split, substring_index,regexp_extract\n",
    "from mmtfPyspark.datasets import pdbToUniProt\n",
    "from ipywidgets import interact, IntSlider, widgets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "spark = SparkSession.builder.appName(\"S-Sulphenylation\").getOrCreate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# dataset = {'BTD':  'https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_002.xlsx', \n",
    "#            'DYn-2':'https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_003.xlsx',\n",
    "#            'PRD':  'https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_004.xlsx',\n",
    "#            'PYD':  'https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_005.xlsx',\n",
    "#            'TD':   'https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_006.xlsx'}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = {'BTD':  ('https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_002.xlsx', '418@C'), \n",
    "           'DYn-2':('https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_003.xlsx', '333@C'),\n",
    "           'PRD':  ('https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_004.xlsx', '333@C'),\n",
    "           'PYD':  ('https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_005.xlsx', '333@C'),\n",
    "           'TD':   ('https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_006.xlsx', '333@C')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = widgets.ToggleButtons(options=dataset.keys(), description='Dataset:', disabled=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "efe2151b6402450ab7b0679ba0d1f1a6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "ToggleButtons(description='Dataset:', options=('BTD', 'DYn-2', 'PRD', 'PYD', 'TD'), value='BTD')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(w)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://pubs.acs.org/doi/suppl/10.1021/jacs.7b01791/suppl_file/ja7b01791_si_002.xlsx 418@C\n"
     ]
    }
   ],
   "source": [
    "dataset[w.value]\n",
    "url = dataset[w.value][0]\n",
    "mod_string = dataset[w.value][1]\n",
    "print(url, mod_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read dataset from supplementary data excel file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset: BTD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Count</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Protein Group</th>\n",
       "      <th>Gene Id</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene Family</th>\n",
       "      <th>Distinct Peptides</th>\n",
       "      <th>Distinct Matches</th>\n",
       "      <th>Filtered Spectra</th>\n",
       "      <th>Description</th>\n",
       "      <th>Modified Sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_008835.5|,NP_001075109.1|</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.221987</td>\n",
       "      <td>68.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>|PRKDC|5591|P78527|DNA-dependent protein kinas...</td>\n",
       "      <td>418@C25,418@C4061,418@C4030,418@C2469,418@C190...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_008835.5|</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.194767</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>|PRKDC|5591|P78527|DNA-dependent protein kinas...</td>\n",
       "      <td>418@C25,418@C223,418@C232,418@C373,418@C458,41...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_001075109.1|</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.249207</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>|PRKDC|5591|P78527|DNA-dependent protein kinas...</td>\n",
       "      <td>418@C25,418@C223,418@C232,418@C373,418@C458,41...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_001447.2|,NP_001104026.1|</td>\n",
       "      <td>249.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.561512</td>\n",
       "      <td>308.0</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>filamin A, alpha</td>\n",
       "      <td>Xq28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>|FLNA|2316|P21333;Q60FE5;Q6NXF2|filamin-A isof...</td>\n",
       "      <td>418@C1018,418@C2152,418@C2160,418@C2535,418@C2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_001447.2|</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.580523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>filamin A, alpha</td>\n",
       "      <td>Xq28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>|FLNA|2316|P21333;Q60FE5;Q6NXF2|filamin-A isof...</td>\n",
       "      <td>418@C53,418@C478,418@C574,418@C1018,418@C1157,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Accession  Cluster  Count   Coverage  Protein Group  \\\n",
       "0  NP_008835.5|,NP_001075109.1|     56.0    2.0   7.221987           68.0   \n",
       "1                  NP_008835.5|      NaN    NaN   7.194767            NaN   \n",
       "2               NP_001075109.1|      NaN    NaN   7.249207            NaN   \n",
       "3  NP_001447.2|,NP_001104026.1|    249.0    2.0  12.561512          308.0   \n",
       "4                  NP_001447.2|      NaN    NaN  12.580523            NaN   \n",
       "\n",
       "  Gene Id                                          Gene Name Chromosome  \\\n",
       "0   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "1   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "2   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "3    FLNA                                   filamin A, alpha       Xq28   \n",
       "4    FLNA                                   filamin A, alpha       Xq28   \n",
       "\n",
       "  Gene Family  Distinct Peptides  Distinct Matches  Filtered Spectra  \\\n",
       "0         NaN               26.0              41.0              73.0   \n",
       "1         NaN                NaN               NaN               NaN   \n",
       "2         NaN                NaN               NaN               NaN   \n",
       "3         NaN               16.0              22.0              41.0   \n",
       "4         NaN                NaN               NaN               NaN   \n",
       "\n",
       "                                         Description  \\\n",
       "0  |PRKDC|5591|P78527|DNA-dependent protein kinas...   \n",
       "1  |PRKDC|5591|P78527|DNA-dependent protein kinas...   \n",
       "2  |PRKDC|5591|P78527|DNA-dependent protein kinas...   \n",
       "3  |FLNA|2316|P21333;Q60FE5;Q6NXF2|filamin-A isof...   \n",
       "4  |FLNA|2316|P21333;Q60FE5;Q6NXF2|filamin-A isof...   \n",
       "\n",
       "                                      Modified Sites  \n",
       "0  418@C25,418@C4061,418@C4030,418@C2469,418@C190...  \n",
       "1  418@C25,418@C223,418@C232,418@C373,418@C458,41...  \n",
       "2  418@C25,418@C223,418@C232,418@C373,418@C458,41...  \n",
       "3  418@C1018,418@C2152,418@C2160,418@C2535,418@C2...  \n",
       "4  418@C53,418@C478,418@C574,418@C1018,418@C1157,...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "req = requests.get(url) # get redirected content\n",
    "df = pd.read_excel(BytesIO(req.content), sheet_name='Protein View')\n",
    "print(\"Dataset:\",w.value)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Standardize representation of protein modification\n",
    "Here we use the following notation for modified residues (amino acid, delta mass), here (C,333)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Count</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Protein Group</th>\n",
       "      <th>Gene Id</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene Family</th>\n",
       "      <th>Distinct Peptides</th>\n",
       "      <th>Distinct Matches</th>\n",
       "      <th>Filtered Spectra</th>\n",
       "      <th>Description</th>\n",
       "      <th>Modified Sites</th>\n",
       "      <th>ptms</th>\n",
       "      <th>modSites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_008835.5|,NP_001075109.1|</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.221987</td>\n",
       "      <td>68.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>|PRKDC|5591|P78527|DNA-dependent protein kinas...</td>\n",
       "      <td>418@C25,418@C4061,418@C4030,418@C2469,418@C190...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>25,4061,4030,2469,1904,3837,3806,1128,1127,234...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_008835.5|</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.194767</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>|PRKDC|5591|P78527|DNA-dependent protein kinas...</td>\n",
       "      <td>418@C25,418@C223,418@C232,418@C373,418@C458,41...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>25,223,232,373,458,457,478,491,1128,1127,1183,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_001075109.1|</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.249207</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>|PRKDC|5591|P78527|DNA-dependent protein kinas...</td>\n",
       "      <td>418@C25,418@C223,418@C232,418@C373,418@C458,41...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>25,223,232,373,458,457,478,491,1128,1127,1183,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_001447.2|,NP_001104026.1|</td>\n",
       "      <td>249.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.561512</td>\n",
       "      <td>308.0</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>filamin A, alpha</td>\n",
       "      <td>Xq28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>|FLNA|2316|P21333;Q60FE5;Q6NXF2|filamin-A isof...</td>\n",
       "      <td>418@C1018,418@C2152,418@C2160,418@C2535,418@C2...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>1018,2152,2160,2535,2543,2191,2199,1157,1453,5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_001447.2|</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.580523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>filamin A, alpha</td>\n",
       "      <td>Xq28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>|FLNA|2316|P21333;Q60FE5;Q6NXF2|filamin-A isof...</td>\n",
       "      <td>418@C53,418@C478,418@C574,418@C1018,418@C1157,...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>53,478,574,1018,1157,1165,1402,1453,1715,1989,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Accession  Cluster  Count   Coverage  Protein Group  \\\n",
       "0  NP_008835.5|,NP_001075109.1|     56.0    2.0   7.221987           68.0   \n",
       "1                  NP_008835.5|      NaN    NaN   7.194767            NaN   \n",
       "2               NP_001075109.1|      NaN    NaN   7.249207            NaN   \n",
       "3  NP_001447.2|,NP_001104026.1|    249.0    2.0  12.561512          308.0   \n",
       "4                  NP_001447.2|      NaN    NaN  12.580523            NaN   \n",
       "\n",
       "  Gene Id                                          Gene Name Chromosome  \\\n",
       "0   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "1   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "2   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "3    FLNA                                   filamin A, alpha       Xq28   \n",
       "4    FLNA                                   filamin A, alpha       Xq28   \n",
       "\n",
       "  Gene Family  Distinct Peptides  Distinct Matches  Filtered Spectra  \\\n",
       "0         NaN               26.0              41.0              73.0   \n",
       "1         NaN                NaN               NaN               NaN   \n",
       "2         NaN                NaN               NaN               NaN   \n",
       "3         NaN               16.0              22.0              41.0   \n",
       "4         NaN                NaN               NaN               NaN   \n",
       "\n",
       "                                         Description  \\\n",
       "0  |PRKDC|5591|P78527|DNA-dependent protein kinas...   \n",
       "1  |PRKDC|5591|P78527|DNA-dependent protein kinas...   \n",
       "2  |PRKDC|5591|P78527|DNA-dependent protein kinas...   \n",
       "3  |FLNA|2316|P21333;Q60FE5;Q6NXF2|filamin-A isof...   \n",
       "4  |FLNA|2316|P21333;Q60FE5;Q6NXF2|filamin-A isof...   \n",
       "\n",
       "                                      Modified Sites ptms  \\\n",
       "0  418@C25,418@C4061,418@C4030,418@C2469,418@C190...  BTD   \n",
       "1  418@C25,418@C223,418@C232,418@C373,418@C458,41...  BTD   \n",
       "2  418@C25,418@C223,418@C232,418@C373,418@C458,41...  BTD   \n",
       "3  418@C1018,418@C2152,418@C2160,418@C2535,418@C2...  BTD   \n",
       "4  418@C53,418@C478,418@C574,418@C1018,418@C1157,...  BTD   \n",
       "\n",
       "                                            modSites  \n",
       "0  25,4061,4030,2469,1904,3837,3806,1128,1127,234...  \n",
       "1  25,223,232,373,458,457,478,491,1128,1127,1183,...  \n",
       "2  25,223,232,373,458,457,478,491,1128,1127,1183,...  \n",
       "3  1018,2152,2160,2535,2543,2191,2199,1157,1453,5...  \n",
       "4  53,478,574,1018,1157,1165,1402,1453,1715,1989,...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.assign(ptms=np.full((df.shape[0], 1), w.value))\n",
    "df['modSites'] = df['Modified Sites'].map(lambda s: str(s).replace(mod_string, ''))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# convert columns to string\n",
    "df[['Gene Family', 'Gene Name', 'Chromosome','Description', 'Modified Sites']] = df[['Gene Family', 'Gene Name', 'Chromosome','Description', 'Modified Sites']].astype(str)\n",
    "ds = spark.createDataFrame(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Count</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Protein Group</th>\n",
       "      <th>Gene Id</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene Family</th>\n",
       "      <th>Distinct Peptides</th>\n",
       "      <th>Distinct Matches</th>\n",
       "      <th>Filtered Spectra</th>\n",
       "      <th>Description</th>\n",
       "      <th>ptms</th>\n",
       "      <th>modSites</th>\n",
       "      <th>unpName</th>\n",
       "      <th>unpId</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_008835.5|,NP_001075109.1|</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.221987</td>\n",
       "      <td>68.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>nan</td>\n",
       "      <td>26.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>[, PRKDC, 5591, P78527, DNA-dependent protein ...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>25</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>P78527</td>\n",
       "      <td>DNA-dependent protein kinase catalytic subunit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_008835.5|,NP_001075109.1|</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.221987</td>\n",
       "      <td>68.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>nan</td>\n",
       "      <td>26.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>[, PRKDC, 5591, P78527, DNA-dependent protein ...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>4061</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>P78527</td>\n",
       "      <td>DNA-dependent protein kinase catalytic subunit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_008835.5|,NP_001075109.1|</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.221987</td>\n",
       "      <td>68.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>nan</td>\n",
       "      <td>26.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>[, PRKDC, 5591, P78527, DNA-dependent protein ...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>4030</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>P78527</td>\n",
       "      <td>DNA-dependent protein kinase catalytic subunit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_008835.5|,NP_001075109.1|</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.221987</td>\n",
       "      <td>68.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>nan</td>\n",
       "      <td>26.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>[, PRKDC, 5591, P78527, DNA-dependent protein ...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>2469</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>P78527</td>\n",
       "      <td>DNA-dependent protein kinase catalytic subunit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_008835.5|,NP_001075109.1|</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.221987</td>\n",
       "      <td>68.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>protein kinase, DNA-activated, catalytic polyp...</td>\n",
       "      <td>8q11</td>\n",
       "      <td>nan</td>\n",
       "      <td>26.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>[, PRKDC, 5591, P78527, DNA-dependent protein ...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>1904</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>P78527</td>\n",
       "      <td>DNA-dependent protein kinase catalytic subunit...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Accession  Cluster  Count  Coverage  Protein Group  \\\n",
       "0  NP_008835.5|,NP_001075109.1|     56.0    2.0  7.221987           68.0   \n",
       "1  NP_008835.5|,NP_001075109.1|     56.0    2.0  7.221987           68.0   \n",
       "2  NP_008835.5|,NP_001075109.1|     56.0    2.0  7.221987           68.0   \n",
       "3  NP_008835.5|,NP_001075109.1|     56.0    2.0  7.221987           68.0   \n",
       "4  NP_008835.5|,NP_001075109.1|     56.0    2.0  7.221987           68.0   \n",
       "\n",
       "  Gene Id                                          Gene Name Chromosome  \\\n",
       "0   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "1   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "2   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "3   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "4   PRKDC  protein kinase, DNA-activated, catalytic polyp...       8q11   \n",
       "\n",
       "  Gene Family  Distinct Peptides  Distinct Matches  Filtered Spectra  \\\n",
       "0         nan               26.0              41.0              73.0   \n",
       "1         nan               26.0              41.0              73.0   \n",
       "2         nan               26.0              41.0              73.0   \n",
       "3         nan               26.0              41.0              73.0   \n",
       "4         nan               26.0              41.0              73.0   \n",
       "\n",
       "                                         Description ptms modSites unpName  \\\n",
       "0  [, PRKDC, 5591, P78527, DNA-dependent protein ...  BTD       25   PRKDC   \n",
       "1  [, PRKDC, 5591, P78527, DNA-dependent protein ...  BTD     4061   PRKDC   \n",
       "2  [, PRKDC, 5591, P78527, DNA-dependent protein ...  BTD     4030   PRKDC   \n",
       "3  [, PRKDC, 5591, P78527, DNA-dependent protein ...  BTD     2469   PRKDC   \n",
       "4  [, PRKDC, 5591, P78527, DNA-dependent protein ...  BTD     1904   PRKDC   \n",
       "\n",
       "    unpId                                               name  \n",
       "0  P78527  DNA-dependent protein kinase catalytic subunit...  \n",
       "1  P78527  DNA-dependent protein kinase catalytic subunit...  \n",
       "2  P78527  DNA-dependent protein kinase catalytic subunit...  \n",
       "3  P78527  DNA-dependent protein kinase catalytic subunit...  \n",
       "4  P78527  DNA-dependent protein kinase catalytic subunit...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds = ds.withColumn('Description', split(ds.Description, \"\\|\"))\n",
    "ds = ds.withColumn(\"unpName\", ds.Description.getItem(1))\n",
    "ds = ds.withColumn(\"unpId\", ds.Description.getItem(3))\n",
    "ds = ds.withColumn(\"name\", ds.Description.getItem(4))\n",
    "ds = ds.withColumn('modSites', split(ds.modSites, \",\"))\n",
    "ds = ds.withColumn('modSites', explode(ds.modSites))\n",
    "ds = ds.drop('Modified Sites')\n",
    "ds.limit(5).toPandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get PDB to UniProt Residue Mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download PDB to UniProt mappings and filter out residues that were not observed in the 3D structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "up = pdbToUniProt.get_cached_residue_mappings().filter(\"pdbResNum IS NOT NULL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "st = up.join(ds, (up.uniprotId == ds.unpId) & (up.uniprotNum == ds.modSites))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "code_folding": [
     0
    ],
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def view_modifications(df, cutoff_distance, *args):\n",
    "\n",
    "    def view3d(show_labels=True,show_bio_assembly=False, show_surface=False, i=0):\n",
    "        pdb_id, chain_id = df.iloc[i]['structureChainId'].split('.')\n",
    "        res_num = df.iloc[i]['pdbResNum']\n",
    "        labels = df.iloc[i]['ptms']\n",
    "        \n",
    "        # print header\n",
    "        print (\"PDB Id: \" + pdb_id + \" chain Id: \" + chain_id)\n",
    "        \n",
    "        # print any specified additional columns from the dataframe\n",
    "        for a in args:\n",
    "            print(a + \": \" + df.iloc[i][a])\n",
    "        \n",
    "        mod_res = {'chain': chain_id, 'resi': res_num}  \n",
    "        \n",
    "        # select neigboring residues by distance\n",
    "        surroundings = {'chain': chain_id, 'resi': res_num, 'byres': True, 'expand': cutoff_distance}\n",
    "        \n",
    "        viewer = py3Dmol.view(query='pdb:' + pdb_id, options={'doAssembly': show_bio_assembly})\n",
    "    \n",
    "        # polymer style\n",
    "        viewer.setStyle({'cartoon': {'color': 'spectrum', 'width': 0.6, 'opacity':0.8}})\n",
    "        # non-polymer style\n",
    "        viewer.setStyle({'hetflag': True}, {'stick':{'radius': 0.3, 'singleBond': False}})\n",
    "        \n",
    "        # style for modifications\n",
    "        viewer.addStyle(surroundings,{'stick':{'colorscheme':'orangeCarbon', 'radius': 0.15}})\n",
    "        viewer.addStyle(mod_res, {'stick':{'colorscheme':'redCarbon', 'radius': 0.4}})\n",
    "        viewer.addStyle(mod_res, {'sphere':{'colorscheme':'gray', 'opacity': 0.7}})\n",
    "        \n",
    "        # set residue labels    \n",
    "        if show_labels:\n",
    "            for residue, label in zip(res_num, labels):\n",
    "                viewer.addLabel(residue + \": \" + label, \\\n",
    "                                {'fontColor':'black', 'fontSize': 8, 'backgroundColor': 'lightgray'}, \\\n",
    "                                {'chain': chain_id, 'resi': residue})\n",
    "\n",
    "        viewer.zoomTo(surroundings)\n",
    "        \n",
    "        if show_surface:\n",
    "            viewer.addSurface(py3Dmol.SES,{'opacity':0.8,'color':'lightblue'})\n",
    "\n",
    "        return viewer.show()\n",
    "       \n",
    "    s_widget = IntSlider(min=0, max=len(df)-1, description='Structure', continuous_update=False)\n",
    "    \n",
    "    return interact(view3d, show_labels=True, show_bio_assembly=False, show_surface=False, i=s_widget)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>structureChainId</th>\n",
       "      <th>pdbResNum</th>\n",
       "      <th>pdbSeqNum</th>\n",
       "      <th>uniprotId</th>\n",
       "      <th>uniprotNum</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Count</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Protein Group</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene Family</th>\n",
       "      <th>Distinct Peptides</th>\n",
       "      <th>Distinct Matches</th>\n",
       "      <th>Filtered Spectra</th>\n",
       "      <th>Description</th>\n",
       "      <th>ptms</th>\n",
       "      <th>modSites</th>\n",
       "      <th>unpName</th>\n",
       "      <th>unpId</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1UCH.A</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>P15374</td>\n",
       "      <td>95</td>\n",
       "      <td>NP_005993.1|</td>\n",
       "      <td>752.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.695652</td>\n",
       "      <td>949.0</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[, UCHL3, 7347, P15374, ubiquitin carboxyl-ter...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>95</td>\n",
       "      <td>UCHL3</td>\n",
       "      <td>P15374</td>\n",
       "      <td>ubiquitin carboxyl-terminal hydrolase isozyme ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1XD3.A</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>P15374</td>\n",
       "      <td>95</td>\n",
       "      <td>NP_005993.1|</td>\n",
       "      <td>752.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.695652</td>\n",
       "      <td>949.0</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[, UCHL3, 7347, P15374, ubiquitin carboxyl-ter...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>95</td>\n",
       "      <td>UCHL3</td>\n",
       "      <td>P15374</td>\n",
       "      <td>ubiquitin carboxyl-terminal hydrolase isozyme ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1XD3.C</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>P15374</td>\n",
       "      <td>95</td>\n",
       "      <td>NP_005993.1|</td>\n",
       "      <td>752.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.695652</td>\n",
       "      <td>949.0</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[, UCHL3, 7347, P15374, ubiquitin carboxyl-ter...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>95</td>\n",
       "      <td>UCHL3</td>\n",
       "      <td>P15374</td>\n",
       "      <td>ubiquitin carboxyl-terminal hydrolase isozyme ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5TTE.E</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>P68036</td>\n",
       "      <td>54</td>\n",
       "      <td>NP_001243284.1|,NP_001243285.1|,NP_003338.1|</td>\n",
       "      <td>629.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>17.466488</td>\n",
       "      <td>784.0</td>\n",
       "      <td>...</td>\n",
       "      <td>UBE2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>[, UBE2L3, 7332, P68036, ubiquitin-conjugating...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>54</td>\n",
       "      <td>UBE2L3</td>\n",
       "      <td>P68036</td>\n",
       "      <td>ubiquitin-conjugating enzyme E2 L3 isoform 1 [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1FBV.C</td>\n",
       "      <td>1054</td>\n",
       "      <td>54</td>\n",
       "      <td>P68036</td>\n",
       "      <td>54</td>\n",
       "      <td>NP_001243284.1|,NP_001243285.1|,NP_003338.1|</td>\n",
       "      <td>629.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>17.466488</td>\n",
       "      <td>784.0</td>\n",
       "      <td>...</td>\n",
       "      <td>UBE2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>[, UBE2L3, 7332, P68036, ubiquitin-conjugating...</td>\n",
       "      <td>BTD</td>\n",
       "      <td>54</td>\n",
       "      <td>UBE2L3</td>\n",
       "      <td>P68036</td>\n",
       "      <td>ubiquitin-conjugating enzyme E2 L3 isoform 1 [...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  structureChainId pdbResNum  pdbSeqNum uniprotId  uniprotNum  \\\n",
       "0           1UCH.A        95         95    P15374          95   \n",
       "1           1XD3.A        95         95    P15374          95   \n",
       "2           1XD3.C        95         95    P15374          95   \n",
       "3           5TTE.E        54         54    P68036          54   \n",
       "4           1FBV.C      1054         54    P68036          54   \n",
       "\n",
       "                                      Accession  Cluster  Count   Coverage  \\\n",
       "0                                  NP_005993.1|    752.0    1.0   8.695652   \n",
       "1                                  NP_005993.1|    752.0    1.0   8.695652   \n",
       "2                                  NP_005993.1|    752.0    1.0   8.695652   \n",
       "3  NP_001243284.1|,NP_001243285.1|,NP_003338.1|    629.0    3.0  17.466488   \n",
       "4  NP_001243284.1|,NP_001243285.1|,NP_003338.1|    629.0    3.0  17.466488   \n",
       "\n",
       "   Protein Group                        ...                          \\\n",
       "0          949.0                        ...                           \n",
       "1          949.0                        ...                           \n",
       "2          949.0                        ...                           \n",
       "3          784.0                        ...                           \n",
       "4          784.0                        ...                           \n",
       "\n",
       "  Gene Family Distinct Peptides Distinct Matches Filtered Spectra  \\\n",
       "0         nan               1.0              1.0              2.0   \n",
       "1         nan               1.0              1.0              2.0   \n",
       "2         nan               1.0              1.0              2.0   \n",
       "3        UBE2               2.0              3.0              6.0   \n",
       "4        UBE2               2.0              3.0              6.0   \n",
       "\n",
       "                                         Description  ptms  modSites unpName  \\\n",
       "0  [, UCHL3, 7347, P15374, ubiquitin carboxyl-ter...   BTD        95   UCHL3   \n",
       "1  [, UCHL3, 7347, P15374, ubiquitin carboxyl-ter...   BTD        95   UCHL3   \n",
       "2  [, UCHL3, 7347, P15374, ubiquitin carboxyl-ter...   BTD        95   UCHL3   \n",
       "3  [, UBE2L3, 7332, P68036, ubiquitin-conjugating...   BTD        54  UBE2L3   \n",
       "4  [, UBE2L3, 7332, P68036, ubiquitin-conjugating...   BTD        54  UBE2L3   \n",
       "\n",
       "    unpId                                               name  \n",
       "0  P15374  ubiquitin carboxyl-terminal hydrolase isozyme ...  \n",
       "1  P15374  ubiquitin carboxyl-terminal hydrolase isozyme ...  \n",
       "2  P15374  ubiquitin carboxyl-terminal hydrolase isozyme ...  \n",
       "3  P68036  ubiquitin-conjugating enzyme E2 L3 isoform 1 [...  \n",
       "4  P68036  ubiquitin-conjugating enzyme E2 L3 isoform 1 [...  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sp = st.toPandas()\n",
    "sp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDB Id: 1UCH chain Id: A\n",
      "uniprotId: P15374\n",
      "unpName: UCHL3\n",
      "name: ubiquitin carboxyl-terminal hydrolase isozyme L3 [Homo sapiens]\n",
      "Accession: NP_005993.1|\n",
      "modSites: 95\n"
     ]
    },
    {
     "data": {
      "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_1579306034501915\"  style=\"position: relative; width: 640px; height: 480px\">\n        <p id=\"3dmolwarning_1579306034501915\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.csb.pitt.edu/build/3Dmol.js');\n}\n\nvar viewer_1579306034501915 = null;\nvar warn = document.getElementById(\"3dmolwarning_1579306034501915\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_1579306034501915 = $3Dmol.createViewer($(\"#3dmolviewer_1579306034501915\"),{backgroundColor:\"white\"});\n$3Dmol.download(\"pdb:1UCH\", viewer_1579306034501915, {\"doAssembly\": false}, function() {\n\tviewer_1579306034501915.setStyle({\"cartoon\": {\"color\": \"spectrum\", \"width\": 0.6, \"opacity\": 0.8}});\n\tviewer_1579306034501915.setStyle({\"hetflag\": true},{\"stick\": {\"radius\": 0.3, \"singleBond\": false}});\n\tviewer_1579306034501915.addStyle({\"chain\": \"A\", \"resi\": \"95\", \"byres\": true, \"expand\": 6},{\"stick\": {\"colorscheme\": \"orangeCarbon\", \"radius\": 0.15}});\n\tviewer_1579306034501915.addStyle({\"chain\": \"A\", \"resi\": \"95\"},{\"stick\": {\"colorscheme\": \"redCarbon\", \"radius\": 0.4}});\n\tviewer_1579306034501915.addStyle({\"chain\": \"A\", \"resi\": \"95\"},{\"sphere\": {\"colorscheme\": \"gray\", \"opacity\": 0.7}});\n\tviewer_1579306034501915.addLabel(\"9: B\",{\"fontColor\": \"black\", \"fontSize\": 8, \"backgroundColor\": \"lightgray\"},{\"chain\": \"A\", \"resi\": \"9\"});\n\tviewer_1579306034501915.addLabel(\"5: T\",{\"fontColor\": \"black\", \"fontSize\": 8, \"backgroundColor\": \"lightgray\"},{\"chain\": \"A\", \"resi\": \"5\"});\n\tviewer_1579306034501915.zoomTo({\"chain\": \"A\", \"resi\": \"95\", \"byres\": true, \"expand\": 6});\nviewer_1579306034501915.render();\n})\n});\n</script>",
      "text/html": [
       "<div id=\"3dmolviewer_1579306034501915\"  style=\"position: relative; width: 640px; height: 480px\">\n",
       "        <p id=\"3dmolwarning_1579306034501915\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
       "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
       "        </div>\n",
       "<script>\n",
       "\n",
       "var loadScriptAsync = function(uri){\n",
       "  return new Promise((resolve, reject) => {\n",
       "    var tag = document.createElement('script');\n",
       "    tag.src = uri;\n",
       "    tag.async = true;\n",
       "    tag.onload = () => {\n",
       "      resolve();\n",
       "    };\n",
       "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
       "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
       "});\n",
       "};\n",
       "\n",
       "if(typeof $3Dmolpromise === 'undefined') {\n",
       "$3Dmolpromise = null;\n",
       "  $3Dmolpromise = loadScriptAsync('https://3dmol.csb.pitt.edu/build/3Dmol.js');\n",
       "}\n",
       "\n",
       "var viewer_1579306034501915 = null;\n",
       "var warn = document.getElementById(\"3dmolwarning_1579306034501915\");\n",
       "if(warn) {\n",
       "    warn.parentNode.removeChild(warn);\n",
       "}\n",
       "$3Dmolpromise.then(function() {\n",
       "viewer_1579306034501915 = $3Dmol.createViewer($(\"#3dmolviewer_1579306034501915\"),{backgroundColor:\"white\"});\n",
       "$3Dmol.download(\"pdb:1UCH\", viewer_1579306034501915, {\"doAssembly\": false}, function() {\n",
       "\tviewer_1579306034501915.setStyle({\"cartoon\": {\"color\": \"spectrum\", \"width\": 0.6, \"opacity\": 0.8}});\n",
       "\tviewer_1579306034501915.setStyle({\"hetflag\": true},{\"stick\": {\"radius\": 0.3, \"singleBond\": false}});\n",
       "\tviewer_1579306034501915.addStyle({\"chain\": \"A\", \"resi\": \"95\", \"byres\": true, \"expand\": 6},{\"stick\": {\"colorscheme\": \"orangeCarbon\", \"radius\": 0.15}});\n",
       "\tviewer_1579306034501915.addStyle({\"chain\": \"A\", \"resi\": \"95\"},{\"stick\": {\"colorscheme\": \"redCarbon\", \"radius\": 0.4}});\n",
       "\tviewer_1579306034501915.addStyle({\"chain\": \"A\", \"resi\": \"95\"},{\"sphere\": {\"colorscheme\": \"gray\", \"opacity\": 0.7}});\n",
       "\tviewer_1579306034501915.addLabel(\"9: B\",{\"fontColor\": \"black\", \"fontSize\": 8, \"backgroundColor\": \"lightgray\"},{\"chain\": \"A\", \"resi\": \"9\"});\n",
       "\tviewer_1579306034501915.addLabel(\"5: T\",{\"fontColor\": \"black\", \"fontSize\": 8, \"backgroundColor\": \"lightgray\"},{\"chain\": \"A\", \"resi\": \"5\"});\n",
       "\tviewer_1579306034501915.zoomTo({\"chain\": \"A\", \"resi\": \"95\", \"byres\": true, \"expand\": 6});\n",
       "viewer_1579306034501915.render();\n",
       "})\n",
       "});\n",
       "</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "view_modifications(sp, 6, 'uniprotId', 'unpName', 'name','Accession','modSites');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "spark.stop()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
